src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -0,0 +1,303 @@
+		values[1] = CStringGetDatum(WaitsForToString(transctionNode->waitsFor));
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * WaitsForToString is only intended for testing and debugging. It gets a
+ * waitsForList and returns the list of transaction nodes' transactionNumber
+ * in a string.
+ */
+static char *
+WaitsForToString(List *waitsFor)
+{
+	StringInfo transactionIdStr = makeStringInfo();
+	ListCell *waitsForCell = NULL;
+
+	foreach(waitsForCell, waitsFor)
+	{
+		TransactionNode *waitingNode = (TransactionNode *) lfirst(waitsForCell);
+
+		if (transactionIdStr->len != 0)
+		{
+			appendStringInfoString(transactionIdStr, ",");
+		}
+
+		appendStringInfo(transactionIdStr, "%ld",
+						 waitingNode->transactionId.transactionNumber);
+	}
+
+	return transactionIdStr->data;
+}
+
+
+/*
+ * FlattenWaitGraph converts a wait graph which is represented as a set of
+ * wait edges between all backends in the Citus cluster, to a set of nodes,
+ * one for each distributed transaction.
+ *
+ * The returned structure is an adjacency list with the following format:
+ * [transactionId] = [transactionNode->waitsFor {list of waiting transaction nodes}]
+ */
+static HTAB *
+FlattenWaitGraph(WaitGraph *waitGraph)
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+	HTAB *distributedTransactionHash = NULL;
+	int edgeIndex = 0;
+	int edgeCount = waitGraph->edgeCount;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(DistributedTransactionId);
+	info.entrysize = sizeof(TransactionNode);
+	info.hash = DistributedTransactionIdHash;
+	info.match = DistributedTransactionIdCompare;
+	info.hcxt = CurrentMemoryContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+
+	distributedTransactionHash =
+		hash_create("distributed deadlock detection", 64, &info, hashFlags);
+
+	for (edgeIndex = 0; edgeIndex < edgeCount; edgeIndex++)
+	{
+		WaitEdge *edge = &waitGraph->edges[edgeIndex];
+		TransactionNode *waitingTransaction = NULL;
+		TransactionNode *blockingTransaction = NULL;
+
+		DistributedTransactionId waitingId = {
+			edge->waitingNodeId,
+			edge->waitingTransactionNum,
+			edge->waitingTransactionStamp
+		};
+
+		DistributedTransactionId blockingId = {
+			edge->blockingNodeId,
+			edge->blockingTransactionNum,
+			edge->blockingTransactionStamp
+		};
+
+		waitingTransaction = GetTransactionNode(distributedTransactionHash, &waitingId);
+		blockingTransaction = GetTransactionNode(distributedTransactionHash, &blockingId);
+
+		waitingTransaction->waitsFor = lappend(waitingTransaction->waitsFor,
+											   blockingTransaction);
+	}
+
+	return distributedTransactionHash;
+}
+
+
+/*
+ * GetTransactionNode searches distributedTransactionHash for the given
+ * given transactionId. If the transaction is not found, the corresponding
+ * transactionNode is initialized to be NULL.
```

Comment:
"the corresponding transactionNode is initialized to be NULL." -> "a new transaction node with the given transaction identifier is added".